### ISSUE TITLE
migrating koin version from 2.1.6 to 3.1.5 broke our generic ViewModel delegation structure

### ISSUE THREAD

**Issue by aytajde** [InsertKoinIO/koin#1323]
**Describe the bug**
We've been using generic type ViewModel in our project and initialising it like below in a BaseFragment

`protected abstract val vmClazz: KClass<VM>
    val viewModel: VM by lazy { getViewModel(vmClazz) { parametersOf(arguments) } }`

_Note: getViewModel method is from the package org.koin.androidx.viewmodel.ext.android_

But, after migrating the koin version to the latest (3.1.5) it doesn't compile anymore and showing the below error:

`Cannot use 'VM' as reified type parameter. Use a class instead.`
`Property delegate must have a 'getValue(BaseFragment<X, X, VM, X>, KProperty<*>)' method. None of the following functions is suitable: 
public inline operator fun <T> Lazy<ViewModel>.getValue(thisRef: Any?, property: KProperty<*>): ViewModel defined in kotlin`

Also, in a  singleton definition when we're trying to use the getProperty() from the package org.koin.core.scope
 file, there is another error in a module file, which is saying,
`Not enough information to infer type variable T`

As I understood, we can't use the generic type ViewModel using the KClass features as it can't infer its type as a generic.

In conclusion, we cannot upgrade our koin version because of the feature misses in a new version.

**Expected behavior**
There are neither a single update info nor the change logs in a documentations. I cannot find any alternatives to accomplish my requirements.
Please, provide these missing features from the previous versions or provide an alternative way to achieve above requirements as we don't want to define ViewModel types explicitly.
**Koin project used and used version (please complete the following information):**
 We are using `    implementation "io.insert-koin:koin-android:2.1.6"
    implementation "io.insert-koin:koin-androidx-viewmodel:2.1.6"`
    
    We want upgrade it to the     implementation   ##'io.insert-koin:koin-android:3.1.5'

**Comment by Vandalko** [comment]
The only workaround (as of version 3.1.6) is to use `java-compat` module.
Honestly, there's no good reason why there's no direct Kotlin counterpart to pass ViewModel class.

**Comment by arnaudgiuliani** [comment]
The ViewModel API is complex to maintain. I'm more thinking to have default APIs and have a kindof DSL to help specify the way you want to use it. We have so much combination of arguments, that it become really hard

**Comment by Zikstar** [comment]
> The only workaround (as of version 3.1.6) is to use `java-compat` module. Honestly, there's no good reason why there's no direct Kotlin equivalent to pass ViewModel class.

Hi, I have this same issue. How do I use the java-compat module. So I just import this dependencyand it should work as before?
    "implementation "io.insert-koin:koin-android-compat:$koinVersion". 
Thats the dependency I found for Java compatiblity in the docs

**Comment by Nemo996** [comment]
Maybe I misunderstood the question, but I solved a similar problem in the following way
abstract class BaseActivity<VM : BaseViewModel, VB : ViewBinding> : AppCompatActivity(),
    AndroidScopeComponent {
    final override val scope : Scope by activityScope()
    protected val viewModel: VM by viewModel(scope,this,getViewModelClass())
Hope this helps you

**Comment by arnaudgiuliani** [comment]
still need some clearer API 👍 Next patch will help

**Comment by arnaudgiuliani** [comment]
Generic access to ViewModel API is available in `koin-android` 3.3.0 👍 
https://github.com/InsertKoinIO/koin/pull/1459
